File: completions-streaming/final/java/src/main/java/com/gloo/streaming/tests/Step2SseParsingTest.java
package com.gloo.streaming.tests;

import com.gloo.streaming.auth.TokenManager;
import com.gloo.streaming.streaming.StreamClient;
import io.github.cdimascio.dotenv.Dotenv;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.util.Map;

/**
 * Streaming Request & SSE Line Parsing Test
 *
 * <p>Validates that:
 * <ul>
 *   <li>makeStreamingRequest() opens a streaming connection</li>
 *   <li>parseSseLine() correctly parses data lines, blank lines, and [DONE]</li>
 * </ul>
 *
 * <p>Usage: mvn -q compile exec:java -Dexec.mainClass=com.gloo.streaming.tests.Step2SseParsingTest
 */
public class Step2SseParsingTest {

    public static void main(String[] args) {
        System.out.println("🧪 Testing: Streaming Request & SSE Line Parsing\n");

        Dotenv dotenv = Dotenv.configure().ignoreIfMissing().load();
        if (dotenv.get("GLOO_CLIENT_ID", "").isBlank()) {
            System.err.println("❌ Missing GLOO_CLIENT_ID — run Step 1 first");
            System.exit(1);
        }

        try {
            String token = TokenManager.ensureValidToken();
            System.out.println("✓ Token obtained\n");

            // Test 1: parseSseLine — blank line
            System.out.println("Test 1: parseSseLine — blank line...");
            Object result = StreamClient.parseSseLine("");
            if (result != null) throw new RuntimeException("Expected null for blank line, got: " + result);
            System.out.println("✓ Blank line → null");

            // Test 2: parseSseLine — non-data line
            System.out.println("Test 2: parseSseLine — non-data line...");
            result = StreamClient.parseSseLine("event: message");
            if (result != null) throw new RuntimeException("Expected null for non-data line, got: " + result);
            System.out.println("✓ Non-data line → null");

            // Test 3: parseSseLine — [DONE] sentinel
            System.out.println("Test 3: parseSseLine — [DONE] sentinel...");
            result = StreamClient.parseSseLine("data: [DONE]");
            if (!"[DONE]".equals(result)) throw new RuntimeException("Expected '[DONE]', got: " + result);
            System.out.println("✓ data: [DONE] → \"[DONE]\"");

            // Test 4: parseSseLine — valid JSON data line
            System.out.println("Test 4: parseSseLine — valid JSON data line...");
            String sample = "data: {\"choices\":[{\"delta\":{\"content\":\"Hello\"},\"finish_reason\":null}]}";
            result = StreamClient.parseSseLine(sample);
            if (!(result instanceof Map)) throw new RuntimeException("Expected Map, got: " + (result == null ? "null" : result.getClass()));
            @SuppressWarnings("unchecked")
            Map<String, Object> chunk = (Map<String, Object>) result;
            @SuppressWarnings("unchecked")
            var choices = (java.util.List<Map<String, Object>>) chunk.get("choices");
            @SuppressWarnings("unchecked")
            var delta = (Map<String, Object>) choices.get(0).get("delta");
            if (!"Hello".equals(delta.get("content"))) throw new RuntimeException("Unexpected content value");
            System.out.println("✓ data: {json} → parsed Map");

            // Test 5: parseSseLine — malformed JSON
            System.out.println("Test 5: parseSseLine — malformed JSON...");
            result = StreamClient.parseSseLine("data: {broken json");
            if (result != null) throw new RuntimeException("Expected null for malformed JSON, got: " + result);
            System.out.println("✓ Malformed JSON → null (gracefully handled)\n");

            // Test 6: Live streaming connection
            System.out.println("Test 6: makeStreamingRequest() — live connection...");
            HttpResponse<java.io.InputStream> response = StreamClient.makeStreamingRequest(
                "Say exactly: 'Stream test OK'", token
            );
            if (response.statusCode() != 200) {
                throw new RuntimeException("Expected 200, got " + response.statusCode());
            }
            System.out.println("✓ Streaming connection opened (status 200)");

            // Test 7: Iterate lines and detect stream termination via finish_reason
            System.out.println("Test 7: Iterating SSE lines and detecting stream termination...");
            int linesProcessed = 0, dataChunks = 0;
            boolean streamTerminated = false;
            String finishReason = null;

            try (BufferedReader reader = new BufferedReader(
                new InputStreamReader(response.body(), StandardCharsets.UTF_8)
            )) {
                String line;
                while ((line = reader.readLine()) != null) {
                    linesProcessed++;
                    Object parsed = StreamClient.parseSseLine(line);
                    if (parsed instanceof Map) {
                        @SuppressWarnings("unchecked")
                        Map<String, Object> chunkMap = (Map<String, Object>) parsed;
                        @SuppressWarnings("unchecked")
                        var lineChoices = (java.util.List<Map<String, Object>>) chunkMap.get("choices");
                        if (lineChoices != null && !lineChoices.isEmpty()) {
                            Object reason = lineChoices.get(0).get("finish_reason");
                            if (reason != null) {
                                streamTerminated = true;
                                finishReason = reason.toString();
                                break;
                            }
                        }
                        dataChunks++;
                    }
                }
            }

            System.out.printf("✓ Processed %d lines, %d data chunks%n", linesProcessed, dataChunks);

            if (!streamTerminated) {
                System.out.println("⚠️  Stream ended without a finish_reason chunk");
            } else {
                System.out.println("✓ Stream terminated cleanly (finish_reason=" + finishReason + ")");
            }

            System.out.println("\n✅ Streaming request and SSE parsing working.");
            System.out.println("   Next: Extracting Token Content\n");

        } catch (Exception e) {
            System.err.println("\n❌ Streaming Request & SSE Parsing Test Failed");
            System.err.println("Error: " + e.getMessage());
            System.err.println("\n💡 Hints:");
            System.err.println("   - Check makeStreamingRequest() sets stream:true in the payload");
            System.err.println("   - Check parseSseLine() strips 'data: ' prefix (substring(6))");
            System.err.println("   - Verify [DONE] check: data.trim().equals(\"[DONE]\")\n");
            System.exit(1);
        }
    }
}

File: completions-streaming/starter/javascript/tests/step2-sse-parsing.test.js
/**
 * Streaming Request & SSE Line Parsing Test
 *
 * Validates that:
 * - makeStreamingRequest() opens a streaming connection
 * - parseSseLine() correctly parses data lines, blank lines, and [DONE]
 *
 * Usage: node tests/step2-sse-parsing.test.js
 */

import "dotenv/config";
import { ensureValidToken } from "../src/auth/tokenManager.js";
import { makeStreamingRequest, parseSseLine } from "../src/streaming/streamClient.js";

async function testStep2() {
  console.log("🧪 Testing: Streaming Request & SSE Line Parsing\n");

  if (!process.env.GLOO_CLIENT_ID) {
    console.error("❌ Missing GLOO_CLIENT_ID — run Step 1 first");
    process.exit(1);
  }

  try {
    const token = await ensureValidToken();
    console.log("✓ Token obtained\n");

    // Test 1: parseSseLine unit tests (no network needed)
    console.log("Test 1: parseSseLine — blank line...");
    let result = parseSseLine("");
    if (result !== null) throw new Error(`Expected null for blank line, got ${JSON.stringify(result)}`);
    console.log("✓ Blank line → null");

    console.log("Test 2: parseSseLine — non-data line...");
    result = parseSseLine("event: message");
    if (result !== null) throw new Error(`Expected null for non-data line, got ${JSON.stringify(result)}`);
    console.log("✓ Non-data line → null");

    console.log("Test 3: parseSseLine — [DONE] sentinel...");
    result = parseSseLine("data: [DONE]");
    if (result !== "[DONE]") throw new Error(`Expected '[DONE]', got ${JSON.stringify(result)}`);
    console.log("✓ data: [DONE] → '[DONE]'");

    console.log("Test 4: parseSseLine — valid JSON data line...");
    const sample = 'data: {"choices":[{"delta":{"content":"Hello"},"finish_reason":null}]}';
    result = parseSseLine(sample);
    if (!result || typeof result !== "object") throw new Error(`Expected object, got ${typeof result}`);
    if (result.choices[0].delta.content !== "Hello") throw new Error("Unexpected content value");
    console.log("✓ data: {json} → parsed object");

    console.log("Test 5: parseSseLine — malformed JSON...");
    result = parseSseLine("data: {broken json");
    if (result !== null) throw new Error(`Expected null for malformed JSON, got ${JSON.stringify(result)}`);
    console.log("✓ Malformed JSON → null (gracefully handled)\n");

    // Test 6: Live streaming connection
    console.log("Test 6: makeStreamingRequest() — live connection...");
    const response = await makeStreamingRequest("Say exactly: 'Stream test OK'", token);

    if (response.status !== 200) {
      throw new Error(`Expected 200, got ${response.status}`);
    }
    console.log("✓ Streaming connection opened (status 200)");

    // Test 7: Iterate chunks and detect stream termination via finish_reason
    console.log("Test 7: Iterating SSE lines and detecting stream termination...");
    const reader = response.body.getReader();
    const decoder = new TextDecoder();
    let buffer = "";
    let linesProcessed = 0;
    let dataChunks = 0;
    let streamTerminated = false;
    let finishReason = null;

    outer: while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split("\n");
      buffer = lines.pop() ?? "";
      for (const line of lines) {
        linesProcessed++;
        const chunk = parseSseLine(line);
        if (chunk && typeof chunk === "object") {
          const reason = chunk.choices?.[0]?.finish_reason;
          if (reason != null) { streamTerminated = true; finishReason = reason; break outer; }
          dataChunks++;
        }
      }
    }
    reader.releaseLock();

    console.log(`✓ Processed ${linesProcessed} lines, ${dataChunks} data chunks`);

    if (!streamTerminated) {
      console.log("⚠️  Stream ended without a finish_reason chunk");
    } else {
      console.log(`✓ Stream terminated cleanly (finish_reason=${finishReason})`);
    }

    console.log("\n✅ Streaming request and SSE parsing working.");
    console.log("   Next: Extracting Token Content\n");
  } catch (err) {
    console.error("\n❌ Streaming Request & SSE Parsing Test Failed");
    console.error(`Error: ${err.message}`);
    console.error("\n💡 Hints:");
    console.error("   - Check makeStreamingRequest() sets stream:true in the payload");
    console.error("   - Check parseSseLine() strips 'data: ' prefix (6 chars with slice(6))");
    console.error("   - Verify [DONE] check: data.trim() === '[DONE]'\n");
    process.exit(1);
  }
}

testStep2();
